### PR TITLE
Hide generated file by erlang.mk

### PR DESCRIPTION
**Pull request by nevar**
#238

### CODE REVIEW

**Comment by nevar** [comment]
Not tested for executable file: kerl, xrefr

**Comment by nevar** [comment]
May be *.o from c_src also hide?

**Comment by essen** [comment]
Why?

**Comment by nevar** [comment]
Don't like add c_src/*.o to .gitignore

**Comment by essen** [comment]
I wouldn't be against an OBJDIR variable, but they're not really temporary files. Intermediate files may be used in more than one way. I notice a few other things in your PR and will comment inline.

File: plugins/c_src.mk [hunk 1]
```diff
@@ -6,7 +6,7 @@
 # Configuration.
 
 C_SRC_DIR ?= $(CURDIR)/c_src
-C_SRC_ENV ?= $(C_SRC_DIR)/env.mk
+C_SRC_ENV ?= $(LOCAL_ERLANG_MK_TMP)/env.mk
```
Comment (essen) [inline plugins/c_src.mk]: This file may be committed, it's not temporary. Really depends on the project.

File: plugins/ci.mk [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 .PHONY: ci ci-setup distclean-kerl
 
-KERL ?= $(CURDIR)/kerl
+KERL ?= $(ERLANG_MK_TMP)/kerl
```
Comment (essen) [inline plugins/ci.mk]: This one is correct, but... does this work and fetch kerl properly?

File: plugins/cover.mk [hunk 3]
```diff
@@ -1,7 +1,7 @@
 # Copyright 2015, Viktor Söderqvist <viktor@zuiderkwast.se>
 # This file is part of erlang.mk and subject to the terms of the ISC License.
 
-COVER_REPORT_DIR = cover
+COVER_REPORT_DIR = $(LOCAL_ERLANG_MK_TMP)/cover
```
Comment (essen) [inline plugins/cover.mk]: The cover directory is an output, not a temporary file. It's similar to the logs directory from Common Test. There's no point in hiding it.

File: plugins/cover.mk [hunk 4]
```diff
@@ -10,15 +10,15 @@ ifdef CT_RUN
 # All modules in 'ebin'
 COVER_MODS = $(notdir $(basename $(call core_ls,ebin/*.beam)))
 
-test-build:: $(TEST_DIR)/ct.cover.spec
+test-build:: $(LOCAL_ERLANG_MK_TMP)/ct.cover.spec
```
Comment (essen) [inline plugins/cover.mk]: ct.cover.spec may be committed and customized.

File: plugins/cover.mk [hunk 5]
```diff
@@ -51,17 +51,17 @@ help::
 
 # Plugin specific targets
 
-COVERDATA = $(filter-out all.coverdata,$(wildcard *.coverdata))
+COVERDATA = $(filter-out $(LOCAL_ERLANG_MK_TMP)/all.coverdata, $(wildcard $(LOCAL_ERLANG_MK_TMP)/*.coverdata))
```
Comment (essen) [inline plugins/cover.mk]: coverdata files are output and may be used afterwards by various tools. It's a bit awkward to have them in the root directory though, would probably be better in cover/

File: plugins/ct.mk [hunk 6]
```diff
@@ -11,6 +11,7 @@ ifneq ($(wildcard $(TEST_DIR)),)
 else
 	CT_SUITES ?=
 endif
+CT_LOGDIR ?= $(LOCAL_ERLANG_MK_TMP)/ct
```
Comment (essen) [inline plugins/ct.mk]: The fact that rebar3 hides the logs/ directory in _build/test/ is my biggest daily annoyance with it. I really don't want to change that. Making it configurable sure, but the default is great.

File: plugins/ct.mk [hunk 7]
```diff
@@ -33,13 +34,12 @@ CT_RUN = ct_run \
...
 	-dir $(TEST_DIR) \
-	-logdir $(CURDIR)/logs
+	-logdir $(CT_LOGDIR)
 
 ifeq ($(CT_SUITES),)
 ct: $(if $(IS_APP),,apps-ct)
 else
-ct: test-build $(if $(IS_APP),,apps-ct)
-	$(verbose) mkdir -p $(CURDIR)/logs/
+ct: test-build $(if $(IS_APP),,apps-ct) | $(CT_LOGDIR)
```
Comment (essen) [inline plugins/ct.mk]: Best to avoid unnecessary changes/make separate commits.

File: plugins/dialyzer.mk [hunk 8]
```diff
@@ -5,7 +5,7 @@
 
 # Configuration.
 
-DIALYZER_PLT ?= $(CURDIR)/.$(PROJECT).plt
+DIALYZER_PLT ?= $(LOCAL_ERLANG_MK_TMP)/.$(PROJECT).plt
```
Comment (essen) [inline plugins/dialyzer.mk]: I'm not convinced the PLT is temporary. It could very well be committed to speed up dialyzing on large projects.

File: plugins/relx.mk [hunk 9]
```diff
@@ -5,7 +5,7 @@
 
 # Configuration.
 
-RELX ?= $(CURDIR)/relx
+RELX ?= $(ERLANG_MK_TMP)/relx
```
Comment (essen) [inline plugins/relx.mk]: Same as kerl.

File: plugins/xref.mk [hunk 10]
```diff
@@ -11,7 +11,7 @@ else
 	XREF_ARGS := -c $(XREF_CONFIG)
 endif
 
-XREFR ?= $(CURDIR)/xrefr
+XREFR ?= $(ERLANG_MK_TMP)/xrefr
```
Comment (essen) [inline plugins/xref.mk]: Same as kerl.

**Comment by essen** [comment]
I have numerous unmerged changes to my local copy. Some are very similar to this, some could benefit from your changes. I will try to do a good mix of both tomorrow.

**Comment by nevar** [comment]
I think we need two configurable variable:
1. ERLANG_MK_TMP for temporary file: xrefr, relx, kerl
2. _WHAT_EVER_ for output: .$(PROJECT).plt, ct, *.coverdata, ct.cover.spec (by default it is $CURDIR)

**Comment by essen** [comment]
So far kerl and relx have been moved to $ERLANG_MK_TMP. xrefr hasn't; I am waiting to take a closer look/document it, it seems pretty basic and perhaps we can do more than that with xref.

Other than that, this remains:

* Move the coverdata files to the cover/ directory
* Review/document cover (and perhaps do something about ct.cover.spec, perhaps not)
* Allow configuring CT logs location

Feel free to submit a separate PR for the CT logs; you already have the work done there, all that's needed is a test and renaming the variable to CT_LOGS_DIR to be more in line with the rest of the variables.

I will be doing the cover stuff soon.

**Comment by essen** [comment]
We should move `ebin/test` into `.erlang.mk/test` also.

**Comment by essen** [comment]
I'll close this. There's most likely a few bits and pieces to improve still, but the bulk of the work was done and we can take each remaining issue independently, if necessary. Thanks!
